### DISCUSSION TITLE
Using Audio Player to play unencoded files

### DISCUSSION THREAD

**Discussion by RDJunc** [pschatzmann/arduino-audio-tools#1288] [Q&A] [answered]
In my application, I am using LittleFS file storage to store short sound clips.  The ESP32 will receive a command to play one of the tracks by getting a track number.  With the emense help that I got previously, I was able to write a small program that could play a designated track using the streaming method.  This demonstrated that a designated file would play but this was just preliminary code.  I am now trying to move forward with the rest of the application.

The sound clips are all stored in headerless files with 16-bit PWM format.

I had originally intended to do it "the hard way" by reading the contents of the file to be played into a buffer and then stream that buffer in the same process as I had used before.  But after studying the AudioPlayer class, that seems like the far better way to go.  It will take in all the names of the files as defined from the source and all I have to do is set the index and initiate the play.  (I assume that the player will have the capabilities as the copier did when streaming when it comes to looping.)

My difficulty is that the AudioPlayer is looking for a decoder.  There is no real decoder so how do I satisfy what the AudioPlayer is looking for.  I played around with Codec16L and may have gotten close but wasn't able to get there.

Thanks as always.

**Comment by pschatzmann** [comment, marked as answer]
Just use a [CopyDecoder](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_copy_decoder.html) which does not change the data

**Reply by RDJunc** [reply]
Well that got me through the first step.  It compiled and loaded.  Nothing played though but that is for the next round of debugging.  Thanks.

**Reply by RDJunc** [reply]
Oh, so close.  The CopyDecoder did the trick when satisfying AudioPlayer.  The .h files are found and (presumably) indexed.  For testing, there are just two.  They do seem to play.  They are there in the background but over taken by shril noise.  I expect that it has something to do with my i2s configuration but there doesn't seem much to work with.  The pins do reflect where they are connected but I could move them if need be.  Unlikely the cause though because the files play just fine using the stream from LittleFS process.

`#include "AudioTools.h"
#include "AudioLibs/AudioSourceLittleFS.h"

const char* startFilePath = "/Media/";
const char* ext = ".h";
AudioSourceLittleFS source(startFilePath, ext);
I2SStream i2s;
CopyDecoder decoder;
AudioPlayer player(source, i2s ,decoder);

void setup() {
	Serial.begin(115200);
	AudioLogger::instance().begin(Serial, AudioLogger::Warning);

	// begin processing
	auto cfg = i2s.defaultConfig();
	cfg.auto_clear = true;
	cfg.sample_rate = 22050;
	cfg.channels = 2;
	cfg.pin_data = 0;
	cfg.pin_bck = 4;
	cfg.pin_ws = 5;

	i2s.begin(cfg);

	// setup player
	//source.setFileFilter("*Bob Dylan*");
	//player.setMetadataCallback(printMetaData);
	player.begin();
}

void loop() {
	player.copy();
}`

[Media.zip](https://github.com/pschatzmann/arduino-audio-tools/files/14157071/Media.zip)

**Comment by RDJunc** [comment]
I made no progress playing the headerless files but wav files play just fine.  However, I have not been able to get the selected file to repeat.  With player.setAutoNext as false, the selected file plays just once but player remains active. How are replays initiated and hopefully without a delay between them.

**Comment by pschatzmann** [comment]
You will need to take care of this yourself by adding some logic. E.g. when you detect the end (e.g. copy returns 0) you can select the next relevant file.

**Comment by RDJunc** [comment]
Done and thank you.

`	if (player.copy() == 0) {
		player.begin();
	}

`
That repeats without the delay and the WAV files play great with no delay.
